From 4318c1e3583ba8499fea09ae21e86c9aa5daac7c Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Thu, 16 Jan 2020 14:01:21 +0800
Subject: [PATCH] PONRTSYS-6336: Fix for ptp packet could not sent out after
 configure rx_los-1pps-pin

    - While linkdown, withing 1 sec, PTP tx skb will be cleared by software,
      At the same time need to clear Xgmac timestamp store fifo
---
 drivers/net/ethernet/lantiq/switch-api/mac/mac_tx_fifo.c |  9 ++++-----
 drivers/net/ethernet/lantiq/switch-api/mac/xgmac_ptp.c   | 14 ++++++++------
 2 files changed, 12 insertions(+), 11 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/mac_tx_fifo.c b/drivers/net/ethernet/lantiq/switch-api/mac/mac_tx_fifo.c
index 02e86d7d70e6..c533e73fb2aa 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/mac_tx_fifo.c
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/mac_tx_fifo.c
@@ -1,5 +1,5 @@
 /******************************************************************************
- *                Copyright (c) 2016, 2017, 2018 Intel Corporation
+ *                Copyright (c) 2016, 2017, 2018, 2019, 2020  Intel Corporation
  *
  *
  * For licensing information, see the file 'LICENSE' in the root folder of
@@ -185,11 +185,10 @@ static int fifo_freeid_get(void *pdev, u8 ttse)
 	struct mac_prv_data *pdata = GET_MAC_PDATA(pdev);
 
 	for (i = START_FIFO; i < MAX_FIFO_ENTRY; i++) {
-		if (!pdata->ts_fifo[i].is_used) {
-			return pdata->ts_fifo[i].rec_id;
-		} else {
+		if (pdata->ts_fifo[i].is_used)
 			fifo_update_hw_clrsts(pdev, &pdata->ts_fifo[i]);
-		}
+		if (!pdata->ts_fifo[i].is_used)
+			return pdata->ts_fifo[i].rec_id;
 	}
 
 	return FIFO_FULL;
diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_ptp.c b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_ptp.c
index 43f94b6a0dc5..52c11875f6b9 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_ptp.c
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_ptp.c
@@ -1,5 +1,5 @@
 /******************************************************************************
- *                Copyright (c) 2016, 2017 Intel Corporation
+ *                Copyright (c) 2016, 2017, 2018, 2019, 2020  Intel Corporation
  *
  *
  * For licensing information, see the file 'LICENSE' in the root folder of
@@ -411,16 +411,18 @@ int xgmac_ptp_tx_work(struct work_struct *work)
 	u64 tstamp;
 	struct skb_shared_hwtstamps shhwtstamp;
 
-	if (!pdata->ptp_tx_skb) {
-		pr_info("No PTP packet transferred out from device\n");
-		return -1;
-	}
-
 	/* check tx tstamp status, if have, read the timestamp and sent to
 	 * stack, otherwise reschedule later
+	 * Get Tx Timestamp need to be done first, since it will clear the
+	 * Xgmac Stored interrupt
 	 */
 	tstamp = pdata->ops.get_tx_tstamp(&pdata->ops);
 
+	if (!pdata->ptp_tx_skb) {
+		pr_debug("No PTP packet transferred out from device\n");
+		return -1;
+	}
+
 	if (!tstamp) {
 		pr_info("Tx tstamp is not captured or ignored for this pkt\n");
 		goto schedule_later;
